hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutAllOperation.java

```diff
@@ -211,12 +214,14 @@ private List toBackupListByRemovingEvictedRecords() {
     protected void writeInternal(ObjectDataOutput out) throws IOException {
         super.writeInternal(out);
         out.writeObject(mapEntries);
+        out.writeBoolean(triggerMapLoader);
```

Comment:
We have a feature called "rolling upgrade" which means that at a certain point, a cluster may consist of members of different minor versions, one minor version apart. In another words, this operation and all other objects that are serialised and deserialised may be sent and received by members of different versions. In the case when this object was sent by a 4.0 member and received by a 4.1 member, this may cause an exception.
This is why we have to introduce additional safeguards. It starts by adding the `Versioned` interface to this class. After that, you can add checks in the `writeInternal` and `readInternal` methods, like:

```java
    @Override
    protected void writeInternal(ObjectDataOutput out) throws IOException {
        super.writeInternal(out);
        out.writeObject(mapEntries);
        if (out.getVersion().isGreaterOrEqual(Versions.V4_1)) {
            out.writeBoolean(triggerMapLoader);
        }
    }

    @Override
    protected void readInternal(ObjectDataInput in) throws IOException {
        super.readInternal(in);
        mapEntries = in.readObject();
        if (in.getVersion().isGreaterOrEqual(Versions.V4_1)) {
            triggerMapLoader = in.readBoolean();
        } else {
            // old behaviour
            triggerMapLoader = true;
        }
    }
```

Same goes for `PutAllPartitionAwareOperationFactory `.

Reply:
That makes sense. I have made this change accordingly. Do we need to test for this specifically?

Reply:
Hm, yes, we have special "compatibility tests" but they are more complicated to write, to run and they don't allow introspecting some internals as they use classloading tricks to separate different instances. We write these tests in our private repository as rolling upgrade is an enterprise feature. So you don't need to be concerned with them.